datastore/providers/postgres_datastore.py

```diff
@@ -0,0 +1,211 @@
+POSTGRES_USERNAME = os.environ.get("POSTGRES_USERNAME", "postgres")
+POSTGRES_PASSWORD = os.environ.get("POSTGRES_PASSWORD", "")
+POSTGRES_DATABASE = os.environ.get("POSTGRES_DATABASE", "pgml_development")
+POSTGRES_TABLENAME = os.environ.get("POSTGRES_TABLENAME", "chatgpt_datastore")
+POSTGRES_SYNCHRONOUS_COMMIT = os.environ.get(
+    "POSTGRES_SYNCHRONOUS_COMMIT", "off"
+).lower()
+
+assert POSTGRES_SYNCHRONOUS_COMMIT in ["on", "off"]
+# OpenAI Ada Embeddings Dimension
+VECTOR_DIMENSION = 1536
+
+
+class PostgresDataStore(DataStore):
+    def __init__(self) -> None:
+        self.conn_params = {
+            "dbname": POSTGRES_DATABASE,
+            "user": POSTGRES_USERNAME,
+            "password": POSTGRES_PASSWORD,
+            "host": POSTGRES_HOST,
+            "port": POSTGRES_PORT,
+        }
+        conn = psycopg.connect(**self.conn_params)
+        # Insert the vector and text into the database
+        create_table = (
+            "CREATE TABLE IF NOT EXISTS %s (doc_id TEXT, chunk_id TEXT, text TEXT, embedding vector(%s), metadata JSONB)"
+            % (POSTGRES_TABLENAME, str(VECTOR_DIMENSION))
+        )
+        cur = conn.cursor()
+        cur.execute(create_table)
+        # Commit the transaction
+        conn.commit()
+        # Close the cursor and the connection
+        cur.close()
+        conn.close()
+
+    async def _upsert(self, chunks: Dict[str, List[DocumentChunk]]) -> List[str]:
+        """
+        Takes in a dict from document id to list of document chunks and inserts them into the index.
+        Return a list of document ids.
+        """
+        # Create a connection
+        conn = psycopg.connect(**self.conn_params)
+        cur = conn.cursor()
+
+        # Initialize a list of ids to return
+        doc_ids: List[str] = []
+
+        # Set synchronous commit
+        cur.execute("SET synchronous_commit = %s" % POSTGRES_SYNCHRONOUS_COMMIT)
+
+        with cur.copy(
+            "COPY %s (doc_id, chunk_id, text, embedding, metadata) FROM STDIN"
+            % POSTGRES_TABLENAME
+        ) as copy:
+            for doc_id, chunk_list in chunks.items():
+                doc_ids.append(doc_id)
+                for chunk in chunk_list:
+                    text = sql.Literal(chunk.text).as_string(conn)
+                    embedding = "[" + ",".join([str(v) for v in chunk.embedding]) + "]"
+
+                    metadata = chunk.metadata.dict()
+                    if "created_at" in list(metadata.keys()):
+                        if metadata["created_at"]:
+                            metadata["created_at"] = to_unix_timestamp(
+                                metadata["created_at"]
+                            )
+                    metadata = json.dumps(metadata)
+
+                    row = (doc_id, chunk.id, text, embedding, metadata)
+                    copy.write_row(row)
+
+        index_statement = (
+            "CREATE INDEX ON %s USING ivfflat (embedding vector_cosine_ops)"
```

Comment:
This should be `CREATE OR REPLACE INDEX` for idempotency, but creating and updating these indexes is so expensive it should definitely be an explicit call.

We should also always use `CONCURRENTLY` for online databases.